File: Eval.py
import sys
import math

#global vars
index = 0 #index of token list
tokens = []
curToken = None

#get user input
def getExpression():
	expression = raw_input("Expression y=? ")
	return expression.lower()

def getX():
	try:
		x = float(input("x? "))
	except NameError:
		sys.exit("Invalid value for x")
	return x

#token class (symbol and numerical values)
class Token:
	def __init__(self,symbol, num):
		self.symbol = symbol
		self.num = num

#print tokens for debugging
def printTokens():
	for t in tokens:
		print t.symbol + " " + str(t.num)

#returns true if string is a valid Token
def isToken(c):
	validTokens = ["(", ")", "+", "-", "*", "/", "^", "cos", 
		"sin", "tan", "ln", "atan", "atan2", ","]
	if c in validTokens:
		return True
	return False

#returns true if string is a number
def isNum(c):
	try:
		float(c)
		return True
	except ValueError:
		return False


#convert expression into tokens
def getTokens(expression):
	global tokens, index, curToken
	i = 0
	while i < len(expression):
		c = expression[i]
		if c.isspace():
			i += 1
			continue
		if i < len(expression) - 4:
			subString = expression[i:i+5]
			if isToken(subString):
				tokens.append(Token(subString, 0))
				i += 5
				continue
		if i < len(expression) - 3:
			subString = expression[i:i+4]
			if isToken(subString):
				tokens.append(Token(subString,0))
				i += 4
				continue
		if i < len(expression)- 2:
			subString = expression[i:i+3]
			if isToken(subString):
				tokens.append(Token(subString,0))
				i += 3
				continue
		if i < len(expression)- 1:
			subString = expression[i:i+2]
			if isToken(subString):
				tokens.append(Token(subString,0))
				i += 2
				continue
		if isToken(c):
			tokens.append(Token(c, 0))
		elif c == "x":
			tokens.append(Token(c, x))
		elif isNum(c):
			s = ""
			count = 0
			subString = expression[i:]
			for char in subString:
				if isNum(char):
					s = s + char
					count += 1
				else:
					break	
			count -=1
			i += count
			tokens.append(Token("int_const", float(s)))
		else:
			return None
		i = i + 1 
	return tokens

#has another token to consume
def hasNextToken():
	global tokens, index, curToken
	if index < len(tokens):
		return True
	return False

#look at next token but doesnt consume it
def peekToken():
	global tokens, index, curToken
	if hasNextToken():
		return tokens[index]
	return None

#consumes token
def consumeToken():
	global tokens, index, curToken
	t = None
	if hasNextToken():
		t = tokens[index]
		index += 1
	return t

#condition for while loop in expr() function
def exprCondition():
	global tokens, index, curToken
	if hasNextToken():
		if peekToken().symbol == "+": return True
		if peekToken().symbol == "-": return True
	return False

#condition for while loop in term() function
def termCondition():
	global tokens, index, curToken
	if hasNextToken():
		if peekToken().symbol == "*": return True
		if peekToken().symbol == "/": return True
	return False

#condition for while loop in exponent() function
def exponentCondtion():
	global tokens, index, curToken
	if hasNextToken():
		if peekToken().symbol == "^": return True
	return False

#recursive decent evaluation
def expr():
	global tokens, index, curToken
	a = term()
	while (exprCondition()):
		s = peekToken().symbol
		curToken = consumeToken()
		b = term()
		if s == "+": a += b
		else: a -= b
	return a

def term():
	global tokens, index, curToken
	a = exponent()
	while(termCondition()):
		s = peekToken().symbol
		curToken = consumeToken()
		b = exponent()
		if s == "*": a *= b
		else: a /= b
	return a

def exponent():
	global tokens, index, curToken
	a = factor()
	while(exponentCondtion()):
		curToken = consumeToken()
		b = factor()
		a = a ** b
	return a

def factor():
	global tokens, index, curToken
	if not hasNextToken():
		sys.exit("Invalid Expression")

	curToken = consumeToken()
	if curToken.symbol == "(":
		a = expr()
		consumeToken() #consume RPARENT or comma
		return a
	elif curToken.symbol == "int_const":
		return curToken.num
	elif curToken.symbol == "sin":
		return math.sin(factor())
	elif curToken.symbol == "cos":
		return math.cos(factor())
	elif curToken.symbol == "tan":
		return math.tan(factor())
	elif curToken.symbol == "atan":
		return math.atan(factor())
	elif curToken.symbol == "atan2":
		a = factor()
		return math.atan2(a, expr())
	elif curToken.symbol == "ln":
		return math.log(factor())
	elif curToken.symbol == "x":
		return curToken.num
	else:
		sys.exit("Invalid Expression")

#main logic
expression = getExpression()
x = getX()
tokens = getTokens(expression)
if tokens == None:
	sys.exit("Invalid token(s)")
curToken = peekToken() #get first token
print expr()

File: Eval.rb

#globals
$index = 0
$tokens = Array.new
$curToken = nil

#get input
def getX()
	print("x? ")
	x = gets.to_f
end

def getExpression()
	print("Expression y=? ")
	expression = gets 
end

class Token
	def initialize(symbol,num)
		@symbol = symbol
		@num = num
	end
	def num
		@num
	end
	def symbol
		@symbol
	end
	def display
		puts "#{@symbol} #{@num}"
	end
end

#print tokens for debugging
def printTokens()
		$tokens.each { |t| t.display } 
end

#returns true if token
def isToken(c)
	validTokens = ["(", ")", "+", "-", "*",
	 "/", "^", "cos", "sin", "tan", "ln", "atan", 
	 "atan2", ","].include?(c)
end

#returns true if number
def isNum(c)
	Integer(c) rescue return false
	return true
end

#convert expression into tokens
def getTokens(expression, x)
	i = 0
	while i < expression.length-1 do 
		c = expression[i]
		foundToken = false
		for j in (4).downto(1)
			if i < expression.length - j
				subString = expression[i..i+j]
				if isToken(subString) 
					$tokens.push(Token.new(subString, 0))
					i += j+1
					foundToken = true
					break	
				end
			end
		end
		foundToken ? next : 0
		if isToken(c)
			$tokens.push(Token.new(c, 0))
		elsif c == "x"
			$tokens.push(Token.new(c, x))
		elsif isNum(c)
			s = String.new
			count = 0
			subString = expression[i..expression.length-1]
			subString.split("").each do |char|
				if isNum(char)
					s << (char)
					count += 1
				else 
					break
				end
			end
			count -= 1
			i += count
			$tokens.push(Token.new("int_const", s.to_f))
		else 
			abort("Invalid Token(s)")
		end
		i += 1
	end 
end

#helper functions
def hasNextToken()
	if $index < $tokens.length
		return true
	end
	return false
end

def peekToken()
	return $tokens[$index]
end

def consumeToken()
	t = $tokens[$index]
	$index += 1
	return t
end

def exprCondition()
	if hasNextToken()
		if peekToken().symbol == "+"
			return true
		end
		if peekToken().symbol == "-"
			return true
		end
	end
	return false
end

def termCondition()
	if hasNextToken()
		if peekToken().symbol == "/"
			return true
		end
		if peekToken().symbol == "*"
			return true
		end
	end
end

def exponentCondition()
	if hasNextToken()
		if peekToken().symbol == "^"
			return true
		end
	end
end

def expr()
	a = term()
	while(exprCondition()) do
		s = peekToken().symbol
		$curToken = consumeToken()
		b = term()
		if s == "+" 
			a += b
		else 
			a -= b
		end 
	end
	return a
end

def term()
	a = exponent()
	while (termCondition()) do
		s = peekToken().symbol
		$curToken = consumeToken()
		b = exponent()
		if s == "*"
			a *= b
		else
			a /= b
		end
	end
	return a
end
def exponent()
	a = factor()
	while (exponentCondition()) do
		s = peekToken().symbol
		$curToken = consumeToken()
		b = factor()
		a = a ** b
	end
	return a
end

def factor()
	if not(hasNextToken())
		abort("Invalid Expression")
	end

	$curToken = consumeToken()
	if $curToken.symbol == "("
		a = expr()
		consumeToken()
		return a
	elsif $curToken.symbol == "int_const"
		return $curToken.num
	elsif $curToken.symbol == "sin"
		 return Math.sin(factor())
	elsif $curToken.symbol == "cos"
		 return Math.cos(factor())
	elsif $curToken.symbol == "tan"
		return Math.tan(factor())
	elsif $curToken.symbol == "atan"
		return Math.atan(factor())
	elsif $curToken.symbol == "atan2"
		a = factor()
		return Math.atan2(a, expr())
	elsif $curToken.symbol == "ln"
		return Math.log(factor())
	elsif $curToken.symbol == "x"
		return $curToken.num
	else
		abort("Invalid Expression")
	end
end
#main logic
expression = getExpression().delete(' ').downcase
x = getX()
getTokens(expression, x)
$curToken = peekToken()
puts expr()


